### PR TITLE
Expose register state

### PR DESCRIPTION
**Pull request by unkadoug**
Define new functions reg_states_iterate and apply_reg_state.

reg_states iterate walks the dwarf code for the procedure identified by the cursor, and invokes a callback for every instruction pointer interval that has a distinct register state, allowing code outside of libunwind to save those register states.

apply_reg_state allows a caller to apply a saved register state to the registers of a cursor. It performs the work of unw_step without the time spent looking up and processing dwarf data.

### CODE REVIEW

File: include/libunwind-common.h.in [hunk 1]
```diff
@@ -126,6 +126,7 @@ typedef struct unw_proc_info
   {
     unw_word_t start_ip;	/* first IP covered by this procedure */
     unw_word_t end_ip;		/* first IP NOT covered by this procedure */
+    unw_word_t last_ip;		/* first IP that could begin another procedure */
```
Comment (djwatson) [inline include/libunwind-common.h.in]: Isn't this an ABI breaking change?

**Comment by djwatson** [comment]
* Your patchset is completely messed up now, with multiple merge comments, empty commits, almost all of them duplicated twice, the first four are already applied.  Please clean it up.
* What's the ultimate goal of this?  A use-case example would be informative.  A unittest would be great. 
* For the documentation generation, use latex2man

**Comment by djwatson** [comment]
(Thanks for breaking it in to individual patches though, much easier to review)

**Comment by unkadoug** [comment]
I agree that the patch set is messed up.  I can try to throw everything away and restart, again.  I don't know what else to do.  It was easier when I could just email patches.

The ultimate goal is to let somebody else (like the hpcrun tool in GitHub.com/HPCToolkit/hpctoolkit) build a mapping of IP address ranges to stack-popping directions, which it does now by disassembling x86 or ppc machine code and inferring stack-popping instructions from that.  I want it to be able to use dwarf data instead, as decoded by libunwind.  I want to be able to do the equivalent of unw_step, without depending on the libunwind cache or repeating the libunwind dwarf decoding more than once.

Re ABI: If you're asking if this changes a publicly visible struct that others may depend on, then yes it does, clearly.  If you have a better idea, I'd welcome it.

**Comment by djwatson** [comment]
> I agree that the patch set is messed up. I can try to throw everything away and restart, again. I don't know what else to do. It was easier when I could just email patches.

Inexperience with git notwithstanding, this should take ~10mins to fix with 'git rebase -i'.  You're still welcome to email patches if you would rather.

Thanks for the example.

I think we should wait on all ABI-breaking changes until a 2.0, so we'll have to either add it a different way, or add it to 2.0.

**Comment by unkadoug** [comment]
I tried rebase.  It wouldn't let me push the result until I had pulled first, so I did.  I can't tell if it helped.

**Comment by unkadoug** [comment]
Destroyed my libunwind repository to get a clean patch set, with a new pull request.

**Pull request closed without merging**